Reading a ConfigMap results in an RBAC problem

luebken opened the pull request:
I'm trying to read a ConfigMap in GuestbookReconciler.Reconcile. I runs fine with `make run` but when I deploy it says it's forbidden:

```
$ make run
...
2021-05-06T23:05:17.478+0200	INFO	controllers.Guestbook	Reconcile called	{"guestbook": "default/guestbook-sample"}
2021-05-06T23:05:17.582+0200	INFO	controllers.Guestbook	Loaded ConfigMap.	{"guestbook": "default/guestbook-sample", "ConfigMap": {"foo":"bar"}}
```

```
$ make docker-build docker-push IMG=luebken/guestbook
$ make deploy IMG=luebken/guestbook

$ kubectl logs guestbook-controller-manager-67989bbbcd-wbfsp -n guestbook-system  manager
...
2021-05-06T21:11:49.422Z	INFO	controllers.Guestbook	Reconcile called	{"guestbook": "default/guestbook-sample"}
E0506 21:11:49.435990       1 reflector.go:127] pkg/mod/k8s.io/client-go@v0.19.2/tools/cache/reflector.go:156: Failed to watch *v1.ConfigMap: failed to list *v1.ConfigMap: configmaps is forbidden: User "system:serviceaccount:guestbook-system:guestbook-controller-manager" cannot list resource "configmaps" in API group "" at the cluster scope
E0506 21:11:50.578774       1 reflector.go:127] pkg/mod/k8s.io/client-go@v0.19.2/tools/cache/reflector.go:156: Failed to watch *v1.ConfigMap: failed to list *v1.ConfigMap: configmaps is forbidden: User "system:serviceaccount:guestbook-system:guestbook-controller-manager" cannot list resource "configmaps" in API group "" at the cluster scope
```

luebken committed:
adding a config

luebken commented:
I've tried adding an kubebuilder rbac definition:
//+kubebuilder:rbac:groups="",namespace=default,resources=configmap,verbs=get;list;watch;create;update;patch;delete
which added a new role `guestbook-manager-role`

And I've tried to bind it to the service account:
`$ kubectl create rolebinding guestbook-controller-manager-rolebinding --role=guestbook-manager-role --serviceaccount=guestbook-system:guestbook-controller-manager -n guestbook-system
`

No success.

jeesmon commented:
@luebken An example that I use

`//+kubebuilder:rbac:groups=core,namespace=system,resources=configmaps,verbs=get;list;watch`

see `groups=core` and `resources=configmaps`

luebken committed:
testing +kubebuilder:rbac

luebken commented:
Thanks @jeesmon. So I've tried both `groups=core` and `groups=""`. In both cases the resulting `roles.yaml` results in 
```
rules:
- apiGroups:
  - ""
 ```
 
 See https://github.com/luebken/guestbook/blob/config-map/controllers/guestbook_controller.go#L42
 
 I've also moved the configmap to system. No luck.
 
 I've also changed the resources to plural: `resources=configmaps`. No luck.

luebken committed:
testing some other suggestions:
* resources plural
* namespace system
* groups core

luebken committed:
moved the configmap into guestbook-system namespac

luebken committed:
added rolebinding (not working)

luebken committed:
(not working)

moved to single namespace=system (guestbook-system)

luebken commented:
By the suggestion I've moved to a single namespace=system (guestbook-system)

Here are the latest role, rolebinding and service account:
https://gist.github.com/luebken/e91de9491c915a99a4e2a45536de8b02

luebken committed:
watching a single namespace

luebken commented:
🥳 Specifying the namespace when setting up the manager was the missing piece:[079b8a2]
